### PR TITLE
[unitaryhack] Add a projector observable

### PR DESCRIPTION
**Pull request by wongwsvincent**
**Context:**
Adding a projector observable to default_qubit device.

**Description of the Change:**
Create a new class ```Projector(Obserable)``` in ```pennylane/ops/qubit.py``` and relevant tests 

**Related GitHub Issues:**
Resolves [issue #1059](https://github.com/PennyLaneAI/pennylane/issues/1059)

### CODE REVIEW

File: pennylane/devices/default_qubit.py
Comment (josh146) [inline pennylane/devices/default_qubit.py:137]: @wongwsvincent, if you have a look at `QubitDevice`, the parent class, it _also_ has a set of supported observables:

https://github.com/PennyLaneAI/pennylane/blob/305279a39a1d70a49ff89b4d09644e62d65e735c/pennylane/_qubit_device.py#L112

This is because, even though `QubitDevice` is an abstract class to be inherited by simulator or hardware devices, simply inheriting it results in automatic support for the listed observables 🙂 This is partly what I alluded to in my comment in the issue --- by adding support for the `Projector` observerable to `QubitDevice`, then _any device that inherits from `QubitDevice`_ will also automatically support this new observable.

File: pennylane/ops/qubit.py
Comment (josh146) [inline pennylane/ops/qubit.py:2964]: @mariaschuld, curious to hear your thoughts on the name? I'm happy with it, although we have (in the past) stuck to the naming convention of `Qubit<thing>` to avoid confusion with `Fock<thing>`. So this would result in naming this `QubitProjector`.

File: pennylane/ops/qubit.py
Comment (josh146) [inline pennylane/ops/qubit.py:3007]: The init method looks perfect! Usage of `qml.math` ensures it will support all machine learning frameworks

File: pennylane/ops/qubit.py
Comment (josh146) [inline pennylane/ops/qubit.py:3017]: An alternative here could be
```suggestion
        idx = np.ravel_multi_index(params[0], [2] * len(params[0]))
```

Since this is purely using NumPy (and avoiding string manipulation), it might be more performant.

Comment (josh146) [reply pennylane/ops/qubit.py:3017]: Although, I might be mistaken. Here are some rough benchmarks I just ran:

```pycon
>>> basis_state = [1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1]
>>> %timeit np.ravel_multi_index(basis_state, [2] * len(basis_state))
13.9 µs ± 516 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit int("".join(str(i) for i in basis_state), 2)
8.33 µs ± 145 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

File: pennylane/ops/qubit.py
Comment (josh146) [inline pennylane/ops/qubit.py:3028]: :+1:

**Review by josh146 — COMMENTED** [review]
Looking good so far @wongwsvincent! My only suggestion is to replicate the change you made in `default_qubit` to `_qubit_device`, so that _any_ inheriting device will also support this observable. 

Then, you can add some integration tests to the device test suite, under `pennylane/devices/tests` :)

File: pennylane/ops/qubit.py
Comment (mariaschuld) [reply pennylane/ops/qubit.py:2964]: I would drop that convention, and rather have `<thing>` and `Fock<thing>`, given that 95% of the ecosystem never interacts with non-qubit methods...

File: .github/CHANGELOG.md
Comment (mariaschuld) [inline .github/CHANGELOG.md:224]: After moving things to `QubitDevice` you could say 
```suggestion
* Added Projector observable, which is available on all devices inheriting from the `QubitDevice` class. [(#1356)](https://github.com/PennyLaneAI/pennylane/pull/1356)
```

File: pennylane/ops/qubit.py
Comment (mariaschuld) [inline pennylane/ops/qubit.py:2983]: nitpicking here, but maybe one could relate `i` to `basis_state`, i.e. that `i` is the integer representation of the bit sequence in `basis_state`, which in turn represents a computational basis state? I always find with these things an example helps, like "for example, the projector |11><11|, or in integer notation |3><3|, is created by `basis_state = [1, 1]`.

File: pennylane/ops/qubit.py
Comment (josh146) [reply pennylane/ops/qubit.py:2964]: Agree 💯

File: pennylane/devices/default_qubit.py
Comment (wongwsvincent) [reply pennylane/devices/default_qubit.py:137]: @josh146 I actually tried adding "Projector" to the list of ```observables``` in _qubit_device.py like I did in default_qubit.py, but it didn't work when I tried it with a ```qiskit.aer``` device. 

After digging a bit more, I now realize that it was because ```observables``` is overwritten in the other package ```pennylane-qiskit```, which doesn't have the "Projector" included.  ```Projector``` is now working with ```qiskit.aer``` if I install the source code of pennylane-qiskit locally and implementing the same change in ```pennylane_qiskit/qiskit_device.py```. 

Does my understanding make sense? If so, I will create a new PR to commit the changes to ```pennylane_qiskit/qiskit_device.py```

File: pennylane/ops/qubit.py
Comment (wongwsvincent) [reply pennylane/ops/qubit.py:3009]: Thank you!

File: pennylane/ops/qubit.py
Comment (wongwsvincent) [reply pennylane/ops/qubit.py:3019]: thanks a lot for checking, @josh146 ! Also, I just checked that the ```ravel_multi_index``` function cannot handle states with 32 or more qubits.
```
>>> basis_state = np.random.randint(2, size=(32))
>>> np.ravel_multi_index(basis_state, [2] * len(basis_state))
ValueError: too many dimensions passed to ravel_multi_index
```
I will keep the original code in.

File: pennylane/ops/qubit.py
Comment (wongwsvincent) [reply pennylane/ops/qubit.py]: Thanks for the suggestion, @mariaschuld 
I have now added a line in the description as the following:
For example, the projector :math:`\ket{11}\bra{11}` , or in integer notation :math:`\ket{3}\bra{3}` is created by ```basis_state=np.array([1, 1])```.

Btw, I see that  I could run ```make docs``` to check the compilation, but is there a way to preview how the description will look on the web documentation?

Comment (wongwsvincent) [reply pennylane/ops/qubit.py]: Please ignore my question, I find the compiled version in ```doc/_build/html/```!

File: pennylane/devices/default_qubit.py
Comment (wongwsvincent) [reply pennylane/devices/default_qubit.py:137]: Extending this discussion to a ```cirq_device```, I find a [projector function in cirq](https://quantumai.google/reference/python/cirq/value/ProductState#projector). It's my first time looking into the cirq library, but adding ```"Projector": CirqOperation(lambda: cirq.ProductState.projector)``` to ```pennylane_cirq/cirq_device.py``` seems to do the job. I will submit a PR for this as well if it's what we want.

File: .github/CHANGELOG.md
Comment (josh146) [inline .github/CHANGELOG.md:224]: Nice 💪

File: pennylane/devices/default_qubit.py
Comment (josh146) [reply pennylane/devices/default_qubit.py:137]: > Does my understanding make sense? If so, I will create a new PR to commit the changes to pennylane_qiskit/qiskit_device.py

Yes this makes perfect sense! Thanks for investigating, it's shame it overwrites this list, otherwise it would have worked out of the box.

A PR contribution is always greatly appreciated 🙂 

> I find a projector function in cirq. It's my first time looking into the cirq library, but adding "Projector": CirqOperation(lambda: cirq.ProductState.projector) to the _observable_map in pennylane_cirq/cirq_device.py seems to do the job. I will submit a PR for this as well if it's what we want.

Yes, this is great! The general idea with the API is as follows:

- A generic, hardware compatible implementation is implemented in `QubitDevice`. This should allow _any_ inheriting device to automatically support these observables in a hardware-compatible manner.

- Some devices, however, might provide their own, more efficient implementations of these expectation values on simulators/hardware, like the Cirq devices. In which cases, these plugins should overwrite to directly use these direct methods 🙂

File: pennylane/ops/qubit.py
Comment (josh146) [inline pennylane/ops/qubit.py:2966]:
```suggestion
    Observable corresponding to the computational basis state projector :math:`P=\ket{i}\bra{i}`.
```

Maybe this wording is slightly more direct?

File: pennylane/ops/qubit.py
Comment (josh146) [reply pennylane/ops/qubit.py:3019]: Perfect, thanks @wongwsvincent!

**Review by josh146 — COMMENTED** [review]
Thanks @wongwsvincent for your thoughtful consideration of the previous round of comments!

I just had a thought, and was wondering what you think. I realized, from looking at the current logic and from [your comment here](https://github.com/PennyLaneAI/pennylane/pull/1356/files#r640140194), that it is not very memory efficient.

For example, this is what `QubitDevice.expval` currently does:

```python
def expval(self, observable, shot_range=None, bin_size=None):
    if self.shots is None:
        # exact expectation value
        eigvals = self._asarray(observable.eigvals, dtype=self.R_DTYPE)
        prob = self.probability(wires=observable.wires)
        return self._dot(eigvals, prob)

    # estimate the ev
    samples = self.sample(observable, shot_range=shot_range, bin_size=bin_size)
    return np.squeeze(np.mean(samples, axis=0))
```

In particular, we are using `dot(eigvals, prob)` to simply index a single element of the probability vector, which is a bit of waste!

Instead, we could simply index directly, which avoids the need to allocate a potentially large `eigvals` array:

```python
def expval(self, observable, shot_range=None, bin_size=None):
    if observable.name == "Projector":
        # branch specifically to handle the projector observable
        idx = int("".join(str(i) for i in bs), 2)
        probs = self.probability(wires=observable.wires, shot_range=shot_range, bin_size=bin_size)
        return probs[idx]

    if self.shots is None:
        # exact expectation value
        eigvals = self._asarray(observable.eigvals, dtype=self.R_DTYPE)
        prob = self.probability(wires=observable.wires)
        return self._dot(eigvals, prob)

    # estimate the ev
    samples = self.sample(observable, shot_range=shot_range, bin_size=bin_size)
    return np.squeeze(np.mean(samples, axis=0))
```

What are your thoughts? This may lead to an improvement when the projector is defined over a large number of wires.

**Comment by wongwsvincent** [comment]
@josh146 , handling ```expval``` differently when Projector is used sounds like a good idea to me. I will add that in.
I am working on the implementation of integration tests under ```pennylane/devices/tests```. I will leave a comment when it's ready for review

File: pennylane/devices/default_qubit.py
Comment (wongwsvincent) [reply pennylane/devices/default_qubit.py:137]: Thanks for confirming, @josh146 !
I have created tickets for PR to pennylane-qiskit ([PR#137](https://github.com/PennyLaneAI/pennylane-qiskit/pull/137)) and pennylane-cirq ([PR#62](https://github.com/PennyLaneAI/pennylane-cirq/pull/62))

Comment (josh146) [reply pennylane/devices/default_qubit.py:137]: Thanks @wongwsvincent! I have seen them come in, but will wait until this PR is merged before beginning the review process there :)

**Comment by wongwsvincent** [comment]
@josh146 , I have included the integration tests for the Projector observable.  The code is ready for another round of review.

File: .github/CHANGELOG.md
Comment (josh146) [reply .github/CHANGELOG.md:238]:
```suggestion
* Added Projector observable, which is available on all devices inheriting from the `QubitDevice` class.
  [(#1356)](https://github.com/PennyLaneAI/pennylane/pull/1356)
```

@wongwsvincent if you would, it would be great to add a minimal code example here to highlight how it is used for anyone reading the changelog after the next release. Have a look at other entries to see how the examples are formatted :)

File: pennylane/devices/tests/test_compare_default_qubit.py
Comment (josh146) [inline pennylane/devices/tests/test_compare_default_qubit.py:71]: great idea to add it to the shared device test suite. This way, it will automatically be tested on all plugin devices.

File: pennylane/devices/tests/test_compare_default_qubit.py
Comment (josh146) [inline pennylane/devices/tests/test_compare_default_qubit.py:84]: How come you are skipping `default.qubit`?

Comment (josh146) [reply pennylane/devices/tests/test_compare_default_qubit.py:84]: Oh nevermind, I just realized because this test compares to `default_qubit`, so it would just be comparing to itself 😆

File: pennylane/_qubit_device.py
Comment (josh146) [inline pennylane/_qubit_device.py:743]: 💪

File: pennylane/devices/tests/test_compare_default_qubit.py
Comment (josh146) [inline pennylane/devices/tests/test_compare_default_qubit.py:103]: Nice combination of basis state tests!

Perhaps it would be best to avoid the `for` loop, and instead parametrize this test using the pytest decorator:

```python
@pytest.mark.parametrize("s", [[0, 0], [0, 1], [1, 0], [1, 1]])
def test_projector_expectation(self, device, s, tol):
```

This has two advantages:

1. PyTest would treat it as 4 independent test cases. Unlike the for loop approach, where one failure will cause the test to exist before all 4 cases are tested.

2. Upon a failure, PyTest will display in the console output the particular value of `s` that caused the failure.

File: pennylane/devices/tests/test_compare_default_qubit.py
Comment (josh146) [inline pennylane/devices/tests/test_compare_default_qubit.py:104]: It might also be good to consider the cases:

- `len(basis_state) < dev.num_wires`
- `qml.Projector(basis_state, wires=[2, 0])` --- that is, cases where the projector is applied to non-consecutive ordered wires. 

These two edge cases often result in more complicated internal logic (e.g., by computing marginal probability distributions), so not testing them can result in bugs not being caught :)

File: pennylane/devices/tests/test_measurements.py
Comment (josh146) [inline pennylane/devices/tests/test_measurements.py:244]: @wongwsvincent this is a great test 🙌 Did you compute the expected projector expectation by hand?

File: pennylane/devices/tests/test_measurements.py
Comment (josh146) [inline pennylane/devices/tests/test_measurements.py:401]: This is wonderful! I didn't even think of testing this edge case, I love it.

I believe this also answers my question above about testing the edge case `len(basis_state) < dev.num_wires` 🙂

File: pennylane/devices/tests/test_compare_default_qubit.py
Comment (josh146) [reply pennylane/devices/tests/test_compare_default_qubit.py:104]: Edit: I see below that `test_projector` has a tensor product that tests the second bullet point, nice!

File: pennylane/devices/tests/test_measurements.py
Comment (josh146) [inline pennylane/devices/tests/test_measurements.py:995]: Amazing test!

Comment (josh146) [reply pennylane/devices/tests/test_measurements.py:995]: This makes me realize though, that

```
var(|b><b|) = E(|b><b|b><b) - E(|b><b|)^2
            = E(|b><b) - E(|b><b|)^2
            = marginal_prob[b] - marginal_prob[b]^2
            = marginal_prob[b] (1-marginal_prob[b])
```

So there is a potential to also speed up `QubitDevice.var`, similarly to how you did with `QubitDevice.expval`:

```python
def var(self, observable, shot_range=None, bin_size=None):
    if observable.name == "Projector":
        # branch specifically to handle the projector observable
        idx = int("".join(str(i) for i in observable.parameters[0]), 2)
        probs = self.probability(
            wires=observable.wires, shot_range=shot_range, bin_size=bin_size
        )
        return probs[idx] - probs[idx] ** 2

    if self.shots is None:
        # exact variance value
        eigvals = self._asarray(observable.eigvals, dtype=self.R_DTYPE)
        prob = self.probability(wires=observable.wires)
        return self._dot((eigvals ** 2), prob) - self._dot(eigvals, prob) ** 2

    # estimate the variance
    samples = self.sample(observable, shot_range=shot_range, bin_size=bin_size)
    return np.squeeze(np.var(samples, axis=0))
```

Having said that, I think you have done more than enough for this PR to be merge-ready, so we could simply make a new GitHub issue, and leave this change for a new PR :)

Comment (josh146) [reply pennylane/devices/tests/test_measurements.py:995]: This also gives me another thought - we should (in another PR!) add to the device API so that you can do

```python
self.basis_state_probability(basis_state, wires)
```

This might be much more memory efficient for some devices, in that the full probability vector is avoided, since we only require a single element of it.

File: pennylane/ops/qubit.py
Comment (josh146) [inline pennylane/ops/qubit.py:2975]:
```suggestion
    For example, the projector :math:`\ket{11}\bra{11}` , or in integer notation :math:`\ket{3}\bra{3}`, is created by ``basis_state=np.array([1, 1])``.
```

File: pennylane/ops/qubit.py
Comment (josh146) [inline pennylane/ops/qubit.py:2985]: docstring looks great :)

**Review by josh146 — CHANGES_REQUESTED** [review]
An amazing PR @wongwsvincent, in particular I was very impressed with both the quality of your tests, and the number of cases you are testing.

I think there is only one thing left before this is merge ready:

- Could you add this new projector to the list of available qubit observables in our reference guide? https://pennylane--1356.org.readthedocs.build/en/1356/introduction/operations.html#qubit-observables. I believe this corresponds to the file `doc/introductions/operations.rst`.

In addition, I left a comment regarding making the `var(Projector)` pipeline more efficient, but that could also be tackled in a separate PR, so I'll leave it up to you.

Thanks @wongwsvincent, especially for continuing to improve the PR throughout our quite full-on code review process :)

File: pennylane/devices/tests/test_compare_default_qubit.py
Comment (wongwsvincent) [reply pennylane/devices/tests/test_compare_default_qubit.py:103]: That totally makes sense! 

This makes me wonder if I should do the same for the other tests in ```test_measurements.py```. 
There are basically four different circuits, so I could define the values of (rotation angles, expected evals, mean and var) at the top, and then use the expected evals, mean and var whenever needed. But this way, only part of the list would be used, and they would be numbers instead of comprehensive formula. So maybe it's better to keep it this way for now.

File: pennylane/devices/tests/test_measurements.py
Comment (wongwsvincent) [reply pennylane/devices/tests/test_measurements.py:244]: Yes, I did all the calculations by hand. I thought of using Mathematica to do the matrix multiplication for me, but I ended up deciding that it shouldn't take too long to calculate by hand. 😅

File: pennylane/devices/tests/test_measurements.py
Comment (wongwsvincent) [reply pennylane/devices/tests/test_measurements.py:995]: Interesting, I agree with that! It does sound like something to do in the next step to be more memory efficient.

**Comment by wongwsvincent** [comment]
@josh146 I feel like too much credit is given to me. I must say the integration tests I implemented was largely inspired by the existing tests for the Hermitian observable 😆  

I have committed the changes as requested:
- removing the for loop in ```test_compare_default_qubit```
- adding this new projector observable to the reference guide

**Comment by wongwsvincent** [comment]
Regarding the improvement of the ```var(Projector)``` pipeline and the implementation of ```self.basis_state_probability(basis_state, wires)```, I feel like I will leave it for another PR in the future.

**Review by josh146 — APPROVED** [review]
Thanks @wongwsvincent! This is ready to go from my side 💯 

Regarding the two changes proposed for subsequent PRs:

- More efficient `QubitDevice.var`: this is something I would like to have before the v0.16 release in a couple of weeks. Let me know if this is something you are interested in helping with, otherwise I'll assign it to a member of the team :)

- Having a better pathway for extracting single basis state probabilities: this likely will require more design work, so maybe best to just open an issue for now.

**Comment by wongwsvincent** [comment]
Thanks for approving the PR and all the help all along the way, @josh146 !
I am happy to help with the ```QubitDevice.var``` function. I will submit a PR when I have a working version.

**Pull request merged**
